Assets/Scripts/Inventory/Inventory.cs

```diff
@@ -2,65 +2,148 @@
+    [System.Serializable]
+    public class ItemSlot
     {
-        this.item = item;
-        this.ItemCount = Count;
+        public ItemSlot(Item item, int Count)
+        {
+            this.item = item;
+            this.ItemCount = Count;
+        }
+        [SerializeField]
+        public Item item;
+        public int ItemCount = 0;
+        public bool Empty()
+        {
+            return ItemCount == 0;
+        }
     }
-    public Item item;
-    public int ItemCount = 0;
-}
-
-public class Inventory : MonoBehaviour
-{
-    public int MaxNumberOfItem = 100;
-    public bool InfiniteItem = false;
 
-    [SerializeField]
-    private List<ItemSlot> Items = new List<ItemSlot>(); // Holds the player's collected items
-    public List<ItemSlot> GetInventoryItemSlots()
+    public class Inventory : MonoBehaviour
     {
-        return Items;
-    }
+        public int MaxNumberOfItem = 100;
+        public bool InfiniteItem = false;
+        public int MaxNumberOfTypes = (int)Item.ItemType.OperationMax; // By default goes to max number of operations
 
-    public void ModifyItemSlot(Item item, int count)
-    {
-        ItemSlot slot = FindItem(item.thisItemType);
+        private void Start()
+        {
+            Items = new List<ItemSlot>(MaxNumberOfTypes);
+            // Regardless of items, we will always have this amount of inventory slots
+            for(int i = 0; i < MaxNumberOfTypes; ++i)
+            {
+                Items.Add(new ItemSlot(null, 0));
+            }
+        }
 
-        if(slot != null) // Item already exists
+        [SerializeField]
+        private List<ItemSlot> Items ; // Holds the player's collected items
+        public List<ItemSlot> GetInventoryItemSlots()
+        {
+            return Items;
+        }
+
+        public void AddNewItemToSlot(ItemSlot slot, Item item, int count)
+        {
+            // If the item slot does not contain an item yet
+            if (slot.item == null)
+            {
+                // ASSMPUTION IS THAT WE ARE ALWAYS ADDING COUNT HERE
+                slot.item = item;
+                slot.ItemCount = count;
+                return;
+            }
+            Debug.LogError("Attempting to add item to a non null item slot");
+        }
+
+        // Given an item slot, change its count
+        public void ModifyItemSlot(ItemSlot slot, int count)
         {
             // Ensure that infinite is not on when we want to modify item count
             if (!this.InfiniteItem)
             {
                 // Ensure we don't go over max
-                if(MaxNumberOfItem < (slot.ItemCount + count))
+                if (MaxNumberOfItem > (slot.ItemCount + count))
                 {
                     // Add count
                     slot.ItemCount += count;
                 }
+                // Ensure we don't go below 0
+                else if (0 >= (slot.ItemCount + count))

```

Comment:
Nit: since you do the calculation of `slot.ItemCount + count` twice, you might as well make a variable at the beginning and reuse that.